### PR TITLE
Tests for the deriveBits operation with extreme values in the 'length' parameter

### PR DESCRIPTION
**Pull request by javifernandez**
There is some inconsistency in the spec regarding the 'length' parameter in the deriveBits operations of different algorithms, such as ECDH, HKDF, PBKDF2 or X25519. This PR adds a few tests to evaluate the behavior of these algorithms with extreme values, like 0, null or undefined. 

There is an ongoing discussion in the issues [322](https://github.com/w3c/webcrypto/issues/322) and [399](https://github.com/w3c/webcrypto/issues/329) about whether the parameter may be optional or accept null values, since it's currently defined as unsigned long. This PR considers the null value as tentative tests while these issues are not solved.

### CODE REVIEW

File: WebCryptoAPI/derive_bits_keys/derived_bits_length.js
Comment (martinthomson) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length.js:6]:
```suggestion
        let testData = algorithms[algorithm];
```

File: WebCryptoAPI/derive_bits_keys/derived_bits_length.js
Comment (martinthomson) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length.js:7]:
```suggestion
        testCases[algorithm].forEach(testParam => {
```

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js
Comment (martinthomson) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js:1]: Do you want to add `length: undefined` cases to this?

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (martinthomson) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:8]: This seems like this 0-length inconsistency with HKDF and PBKDF2 should be fixed at the same time. Pick one.

File: WebCryptoAPI/derive_bits_keys/derived_bits_length.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length.js:6]: Donw

File: WebCryptoAPI/derive_bits_keys/derived_bits_length.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length.js:7]: Done

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js:1]: If I've understood correctly, the agreement we have in [PR#345](https://github.com/w3c/webcrypto/pull/345) is to coerce ```undefined``` to ```null```, since this is what I believe the proposed expression does: ```optional unsigned long? length = null```

Additionally, the ```undefined``` length concept is not explicitly mentioned in the spec, so unless there are good reasons, I'd leave it aside of these test for now.

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:8]: Do you mean those test cases are redundant ? I think there is still value on being exhaustive on the algorithms included in the subtests. Additionally, the undefined expected value means that the test will check that an OperationError exception is thrown, and not a different one.

Comment (martinthomson) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:8]: I mean that the tests produce a different outcome, when the behaviour we're requesting is essentially the same ("generate 0 bytes from your KDF please").

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js
Comment (martinthomson) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js:1]: The coercion happens, but you aren't testing that it has been implemented correctly.

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:8]: Yeah, indeed. I was confused by the fact that in the TR version we were not considering non-multiple of 8 cases as OperatonError. It's true that we have fixed this already in the last spec draft. 

I'll fix this by using the 'undefined' value for the expected result, so we assert this case throws OperationError for both algorithms.

**Comment by javifernandez** [comment]
@twiss @martinthomson if there isn't anything else to change on the new tests, could you please review the PR so that I can merge it ? Thanks.

**Comment by twiss** [comment]
Hey :wave: Apologies for the delayed response.

I think the new (non-tentative) tests are correct in that they match the spec, however, for ECDH and X25519, I'm not 100% sure whether we should start testing that passing 0 leads to an empty array at this time, before a decision has been reached on how to handle the parameter.

One of the proposals I made in https://github.com/w3c/webcrypto/pull/345#issuecomment-1683463216 was to ignore the `length` parameter entirely. If we'd go with that option, even the non-tentative tests would have to change again.

Obviously, we can always do so, and in a sense that reflects the fact that it would be a backwards-incompatible change (and hopefully the counters from Chromium will inform us whether that's possible). But, perhaps it's easier to wait for that? (I'll also try to ping the people in that thread early next week, to try to get to a decision / conclusion.)

Also because, Safari currently returns the entire value when passing 0, and on the (admittedly very unlikely) off-chance that someone relies on this, returning an empty array could lead to a security issue. So if we do end up going with that option, perhaps we don't want to encourage Safari to change the behavior to returning an empty array now (and then back later).

**Comment by javifernandez** [comment]
> Hey 👋 Apologies for the delayed response.
> 

Don't worry, I'm the one that should apologize for being so insistent when asking for the review. I was waiting for this tests to land this Chrome [CL](https://chromium-review.googlesource.com/c/chromium/src/+/5013612) to make the X25519 implementation matching the current (and being discussed) spec, so that zero length returns an empty string, updating the WPT accordingly, since 
we have a few cases with 'null' that seem wrong.

> I think the new (non-tentative) tests are correct in that they match the spec, however, for ECDH and X25519, I'm not 100% sure whether we should start testing that passing 0 leads to an empty array at this time, before a decision has been reached on how to handle the parameter.

We are now in a weird situation, where there are tests that use 'null' when it's not allowed in the current spec, with the fill array as expected result; the only way the implementations can do that is to make '0' equivalent to 'null' (this is Chrome's current behavior) which goes against the spec. 

The CL I mentioned before tries to make the X25519 implementation to mach the current spec, returning an empty array for a '0' length; it causes the mentioned test to fail, though. With this PR I tried to fix also the mess we have in the tests, but I understand your concerns about it; I'm just afraid the agreement on the spec issue takes too long. 

Unless anybody has good arguments against, I think landing the mentioned CL to that Chrome's X25519 matches the current spec (0 returns an empty string) is the best option now, although I really wanted to land it with the proper testing coverage for the 'zero' length case; if there is no agreement, I may have remove the new test cases there for that case.

> 
> One of the proposals I made in [w3c/webcrypto#345 (comment)](https://github.com/w3c/webcrypto/pull/345#issuecomment-1683463216) was to ignore the `length` parameter entirely. If we'd go with that option, even the non-tentative tests would have to change again.
> 
> Obviously, we can always do so, and in a sense that reflects the fact that it would be a backwards-incompatible change (and hopefully the counters from Chromium will inform us whether that's possible). But, perhaps it's easier to wait for that? (I'll also try to ping the people in that thread early next week, to try to get to a decision / conclusion.)
> 
> Also because, Safari currently returns the entire value when passing 0, and on the (admittedly very unlikely) off-chance that someone relies on this, returning an empty array could lead to a security issue. So if we do end up going with that option, perhaps we don't want to encourage Safari to change the behavior to returning an empty array now (and then back later).

In the case of Chrome, the X25519 algorithm is still behind an experimental runtime flag, so I don't think it's a problem to change the implementation and align it with the current spec (zero length returns an empty array), but I guess we can wait until the spec issue is resolved to merge the PR with the definitive tests.

**Comment by twiss** [comment]
Yeah, I understand. I'm also not really worried about adding tests for X25519 (tentative or otherwise) that need to be changed later, but I'm more hesitant about doing that for ECDH.

But, if you don't want to wait for the spec decision, perhaps we can also mark the tests for 0 as being tenative, and merge them like that, for now?

**Comment by javifernandez** [comment]
> Yeah, I understand. I'm also not really worried about adding tests for X25519 (tentative or otherwise) that need to be changed later, but I'm more hesitant about doing that for ECDH.

Makes sense.

> 
> But, if you don't want to wait for the spec decision, perhaps we can also mark the tests for 0 as being tenative, and merge them like that, for now?

The CL changed the Chrome's X25519 implementation to change its behavior for the '0' length case that's why I wanted to land this PR under the assumption that there was more agreement on that case, leaving as tentative the 'null' and 'undefined' ones. If we want to mark all as tentative, perhaps I can land the CL without new tests and  can wait until we solve the issue in the spec, at least for the '0' case.

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js
Comment (panva) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js:1]: Can you please rename the file to `WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases_tentative.js` or similar? Like so the file gets picked up by node's WPT runner as a test-containing file.

Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.tentative.js:1]: Sure.

**Comment by panva** [comment]
Can we update this for the merged state of https://github.com/w3c/webcrypto/pull/345?

**Comment by panva** [comment]
@javifernandez Can you update this for the merged state of https://github.com/w3c/webcrypto/pull/345?

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (panva) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:32]:
```suggestion
        {length: undefined, expected: algorithms["X25519"].derivation},
        {length: "omitted", expected: algorithms["X25519"].derivation }, // default value is null
```

**Review by panva — CHANGES_REQUESTED** [review]

**Comment by panva** [comment]
I don't believe all -1 tests are correct.

https://webidl.spec.whatwg.org/#abstract-opdef-converttoint

```
If x < lowerBound or x > upperBound, thenthrow a TypeError.
```

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (panva) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:4]: This is also incorrect. https://w3c.github.io/webcrypto/#hkdf-operations

```
If length is null or zero, or is not a multiple of 8, then [throw](https://webidl.spec.whatwg.org/#dfn-throw) an [OperationError](https://webidl.spec.whatwg.org/#operationerror).
```

**Review by panva — CHANGES_REQUESTED** [review]

File: WebCryptoAPI/derive_bits_keys/derived_bits_length.js
Comment (panva) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length.js:16]:
```suggestion
                    if (testParam.length === "omitted") {
```

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (javifernandez) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:4]: Umm, you're right. Then, I have doubts that test case makes sense for this test. What do you think ?

Comment (panva) [reply WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:4]: It doesn't seem appropriate, I think it makes sense to test explicit null as `length` and omitted `length`.

**Review by panva — APPROVED** [review]

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (twiss) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:4]: Pedantic nit: 0 is a multiple of 8, but it's disallowed separately.
```suggestion
        {length: 0, expected: undefined}, // should throw an exception
        {length: 1, expected: undefined}, // should throw an exception (not a multiple of 8)
```

File: WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js
Comment (twiss) [inline WebCryptoAPI/derive_bits_keys/derived_bits_length_testcases.js:11]:
```suggestion
        {length: 0, expected: undefined}, // should throw an exception
        {length: 1, expected: undefined}, // should throw an exception (not a multiple of 8)
```

**Review by twiss — APPROVED** [review]
:+1: Thanks!

**Pull request merged**
